File: csharp/src/Training.Module8/BoyScout/Before/ReleaseSummaryFormatter.cs
namespace Training.Module8.BoyScout.Before;

public sealed class ReleaseSummaryFormatter
{
    public string Format(
        string releaseId,
        IReadOnlyList<DeploymentResult> results)
    {
        ArgumentNullException.ThrowIfNull(releaseId, "releaseId");
        ArgumentNullException.ThrowIfNull(results, "results");
        if (string.IsNullOrWhiteSpace(releaseId))
        {
            throw new ArgumentException(
                "releaseId must not be blank");
        }
        if (results.Count == 0)
        {
            throw new ArgumentException(
                "results must not be empty");
        }

        string s = "Release " + releaseId.Trim() + "\n";
        int n = 0;
        foreach (DeploymentResult r in results)
        {
            ArgumentNullException.ThrowIfNull(r, "result");
            if (r.Status == DeploymentStatus.Success)
            {
                s = s + "[OK] " + r.Environment + ": "
                    + r.Description + "\n";
                n++;
            }
            else
            {
                s = s + "[ERROR] " + r.Environment + ": "
                    + r.Description + "\n";
            }
        }
        return s + "Successful: " + n + "/" + results.Count;
    }
}

File: typescript/src/module8/boyscout/before/ReleaseSummaryFormatter.ts
import { IllegalArgumentError } from '../../../shared/errors.js';
import { requireNonNull } from '../../../shared/requireNonNull.js';
import type { DeploymentResult } from '../DeploymentResult.js';
import { DeploymentStatus } from '../DeploymentStatus.js';

export class ReleaseSummaryFormatter {
  format(releaseId: string, results: readonly DeploymentResult[]): string {
    requireNonNull(releaseId, 'releaseId');
    requireNonNull(results, 'results');
    if (releaseId.trim().length === 0) {
      throw new IllegalArgumentError('releaseId must not be blank');
    }
    if (results.length === 0) {
      throw new IllegalArgumentError('results must not be empty');
    }

    let s = 'Release ' + releaseId.trim() + '\n';
    let n = 0;
    for (const r of results) {
      requireNonNull(r, 'result');
      if (r.status === DeploymentStatus.SUCCESS) {
        s = s + '[OK] ' + r.environment + ': ' + r.description + '\n';
        n++;
      } else {
        s = s + '[ERROR] ' + r.environment + ': ' + r.description + '\n';
      }
    }
    return s + 'Successful: ' + n + '/' + results.length;
  }
}
